tests/models/test_selector.py

```diff
@@ -0,0 +1,113 @@
+                    True: {
+                        Rule("event_type", LazyModelField("context__path")): {
+                            True: ServerEventModel,
+                            False: None,
+                        },
+                    },
+                    False: None,
+                }
+            },
+        ),
+        (
+            {
+                PageCloseBrowserEventModel: [
+                    Rule("event_source", "browser"),
+                    Rule("event_type", "page_close"),
+                ],
+                ServerEventModel: [
+                    Rule("event_source", "server"),
+                    Rule("event_type", LazyModelField("context__path")),
+                ],
+            },
+            {
+                Rule("event_source", "browser"): {
+                    True: {
+                        Rule("event_type", "page_close"): {
+                            True: PageCloseBrowserEventModel,
+                            False: None,
+                        }
+                    },
+                    False: {
+                        Rule("event_source", "server"): {
+                            True: {
+                                Rule("event_type", LazyModelField("context__path")): {
+                                    True: ServerEventModel,
+                                    False: None,
+                                }
+                            },
+                            False: None,
+                        }
+                    },
+                }
+            },
+        ),
+    ],
+)
+def test_models_selector_model_selector_decision_tree(model_rules, decision_tree):
+    """Given `model_rules` the ModelSelector should create the expected decision tree."""
+
+    model_selector = ModelSelector()
+    assert model_selector.get_decision_tree(model_rules) == decision_tree
+
+
+@pytest.mark.parametrize(
+    "event,model",
+    [
+        (ServerEventFactory(), ServerEventModel),
+        (PageCloseBrowserEventFactory(), PageCloseBrowserEventModel),
+        ({"invalid": "event"}, None),
+    ],
+)
+def test_models_selector_model_selector_get_model(event, model):
+    """Given an event the ModelSelector.get_model method should return the corresponding model."""
+
+    event = json.loads(event.json()) if not isinstance(event, dict) else event
+    assert ModelSelector().get_model(event) is model
+
+
+def test_models_selector_model_selector_model_rules():
+    """Checks that no model_rules set is included in another model_rules set."""
+
+    model_rules = ModelSelector().model_rules
+    for model, rules in model_rules.items():
+        rules_set = set(rules)
+        other_model_rules = filter(lambda x, m=model: x[0] != m, model_rules.items())
+        for other_model, other_rules in other_model_rules:
+            if rules_set.issubset(set(other_rules)):
+                pytest.fail(f"The rules of {model} are a subset of {other_model}")
```

Comment:
That's a great idea, but as discussed IRL it should be checked while instantiating the decision tree.

Reply:
Thanks) You are right) Although it costs a bit of processing at the instantiation it's worth to check the decision tree while building it.